### ISSUE TITLE
Adding a git source for a transitive dependency that is already locked does not update the remote version

### ISSUE THREAD

**Issue by Earlopain** [rubygems/rubygems#8238]
### Describe the problem as clearly as you can

Start out with the following files:

```rb
# Gemfile
source "https://rubygems.org"

gem "activesupport"
```

```rb
# Gemfile.lock
GEM
  remote: https://rubygems.org/
  specs:
    activesupport (8.0.0)
      base64
      benchmark (>= 0.3)
      bigdecimal
      concurrent-ruby (~> 1.0, >= 1.3.1)
      connection_pool (>= 2.2.5)
      drb
      i18n (>= 1.6, < 2)
      logger (>= 1.4.2)
      minitest (>= 5.1)
      securerandom (>= 0.3)
      tzinfo (~> 2.0, >= 2.0.5)
      uri (>= 0.13.1)
    base64 (0.2.0)
    benchmark (0.4.0)
    bigdecimal (3.1.8)
    concurrent-ruby (1.3.4)
    connection_pool (2.4.1)
    drb (2.2.1)
    i18n (1.14.6)
      concurrent-ruby (~> 1.0)
    logger (1.6.1)
    minitest (5.25.1)
    securerandom (0.3.1)
    tzinfo (2.0.6)
      concurrent-ruby (~> 1.0)
    uri (1.0.1)

PLATFORMS
  ruby
  x86_64-linux

DEPENDENCIES
  activesupport

BUNDLED WITH
   2.6.0.dev
```

Important to note here is that `securerandom` is at version `0.3.1`, which is not the latest.

Now, modify the gemfile to the following:
```rb
source "https://rubygems.org"

gem "activesupport"
gem "securerandom", github: "ruby/securerandom" # This is new
```

Run `bundle lock`. The lockfile now contains this:
```rb
GIT
  remote: https://github.com/ruby/securerandom.git
  revision: 9703d96a76ee72a278558a474d69cc0eb0bc0d6e
  specs:
    securerandom (0.3.1)

GEM
  remote: https://rubygems.org/
  specs:
    activesupport (8.0.0)
      base64
      benchmark (>= 0.3)
      bigdecimal
      concurrent-ruby (~> 1.0, >= 1.3.1)
      connection_pool (>= 2.2.5)
      drb
      i18n (>= 1.6, < 2)
      logger (>= 1.4.2)
      minitest (>= 5.1)
      securerandom (>= 0.3)
      tzinfo (~> 2.0, >= 2.0.5)
      uri (>= 0.13.1)
    base64 (0.2.0)
    benchmark (0.4.0)
    bigdecimal (3.1.8)
    concurrent-ruby (1.3.4)
    connection_pool (2.4.1)
    drb (2.2.1)
    i18n (1.14.6)
      concurrent-ruby (~> 1.0)
    logger (1.6.1)
    minitest (5.25.1)
    tzinfo (2.0.6)
      concurrent-ruby (~> 1.0)
    uri (1.0.1)

PLATFORMS
  ruby
  x86_64-linux

DEPENDENCIES
  activesupport
  securerandom!

BUNDLED WITH
   2.6.0.dev
```
Which is incorrect. The remote got bumped to `0.3.2` but the git spec still states `0.3.1`. During install, it will throw the following error:
```sh
$ bundle install
Fetching gem metadata from https://rubygems.org/........
Your bundle is locked to securerandom (0.3.1) from https://github.com/ruby/securerandom.git (at master@9703d96), but that version can no longer be
found in that source. That means the author of securerandom (0.3.1) has removed it. You'll need to update your bundle to a version other than
securerandom (0.3.1) that hasn't been removed in order to install.
```

I don't get this error if `securerandom` is already part of my bundle via a direct dependency. In that case, the version is getting correctly updated.

### Did you try upgrading rubygems & bundler?

I'm fairly close to the current master branch (2024-11-05)

## Environment

```
Bundler       2.6.0.dev
  Platforms   ruby, x86_64-linux
Ruby          3.4.0p-1 (2024-11-05 revision ed06f018bdffe9bb7f8bdbf15fa5a727e402bfe9) [x86_64-linux]
  Full Path   /home/earlopain/.rbenv/versions/3.4-dev/bin/ruby
  Config Dir  /home/earlopain/.rbenv/versions/3.4-dev/etc
RubyGems      3.6.0.dev
  Gem Home    /home/earlopain/.rbenv/versions/3.4-dev/lib/ruby/gems/3.4.0+0
  Gem Path    /home/earlopain/.gem/ruby/3.4.0+0:/home/earlopain/.rbenv/versions/3.4-dev/lib/ruby/gems/3.4.0+0
  User Home   /home/earlopain
  User Path   /home/earlopain/.gem/ruby/3.4.0+0
  Bin Dir     /home/earlopain/.rbenv/versions/3.4-dev/bin
Tools         
  Git         2.47.0
  RVM         not installed
  rbenv       rbenv 1.2.0
  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2024-11-13
Git SHA           unknown
Released Version  false
```

## Bundler settings

```
gem.changelog
  Set for the current user (/home/earlopain/.bundle/config): false
gem.ci
  Set for the current user (/home/earlopain/.bundle/config): "github"
gem.coc
  Set for the current user (/home/earlopain/.bundle/config): false
gem.linter
  Set for the current user (/home/earlopain/.bundle/config): "rubocop"
gem.mit
  Set for the current user (/home/earlopain/.bundle/config): true
gem.test
  Set for the current user (/home/earlopain/.bundle/config): "rspec"
```

**Comment by deivid-rodriguez** [comment]
Thanks for reporting, working on a fix at #8269. It should be ready provided CI passes.

**Comment by Earlopain** [comment]
Cool! I tested out your PR and it seems to work fine.
